### ISSUE TITLE
JSON Schema generation?

### ISSUE THREAD

**Issue by mtiller** [woutervh-/typescript-is#40]
In addition to the current functionality, it would be very nice to have JSON schema generation as part of the transform.  This would provide a means of communicating types to other non-typescript systems.  For example, the ability to perform client side validation of data by pushing out JSON schema to the clients so they could do their own local validation prior to submitting data to an HTTP API, for example.

**Comment by woutervh-** [comment]
Hi @mtiller 

Indeed it's a nice way to generate JSON schema consistent with your TypeScript interfaces.
There are already some projects out there that can do this:
- https://github.com/YousefED/typescript-json-schema
- https://github.com/lbovet/typson
- https://github.com/vega/ts-json-schema-generator

Hope this helps :)

**Comment by sparebytes** [comment]
Inspired by this project, I wrote my own transformation plugin that actually generates JSON schema and passes it through AJV for validation. It doesn't emit separate json files but emits it into the javascript. One caveat is that right now each validation function has a full-blown copy of the schema needed. EG if you have validation functions for both `interface Animal` and `interface Cat extends Animal`, the `Cat` validator will contain duplicate JSON Schema code from `Animal`.

Anyway I'm not attesting to whether my project is any better or worse than this one. The reason I went this route was to make sure my own esoteric use cases are supported. Also by using `ts-json-schema-generator` it minimizes the code for traversing the the Typescript AST inside my own project.

https://github.com/sparebytes/typesmith
